### DISCUSSION TITLE
Memory Leak for Pekko Shardedc cluster

### DISCUSSION THREAD

**Discussion by Susmit07** [apache/pekko-connectors#885] [General]
Hellp Team

We were tying large file upoad 15-20gb using HDFS as source and S3 as sink with  S3 multipartApi, but I am seeing leaking resources

```
One instance of java.util.zip.ZipFile$Source loaded by <system class loader> occupies 4,94,11,544 (55.80%) bytes. The instance is referenced by org.apache.pekko.dispatch.MonitorableThreadFactory$PekkoForkJoinWorkerThread @ 0x6c33b5548 cef-datasync-downloader-pekko.actor.default-dispatcher-6, loaded by jdk.internal.loader.ClassLoaders$AppClassLoader @ 0x6c34cda68.
```

<img width="1070" alt="image" src="https://github.com/user-attachments/assets/704ad329-00a3-4590-9ae9-782493a6e1bf">

<img width="826" alt="image" src="https://github.com/user-attachments/assets/79ec2844-a708-4d44-9dc3-b9244c381a61">

<img width="826" alt="image" src="https://github.com/user-attachments/assets/e68c3651-54d7-4528-87db-424fc1f3437e">

<img width="826" alt="image" src="https://github.com/user-attachments/assets/1ec8a728-9c53-4060-87a0-ffbcea42052c">

ZiFileSource is tha major culprit here

Any suggestion elcome

**Comment by pjfanning** [comment]
A memory leak occurs when an app uses memory but does not release it when it is finished with that data. You have provided no evidence that the memory was leaked - just that a lot of memory was used. java.util.zip.ZipFile is a very memory intensive way to read zip data.
There is a java.util.zip.ZipInputStream class that can be used to stream through the zip data instead. It may be possible to support this approach but I haven't read the code yet.

**Reply by pjfanning** [reply]
Looking at this closer, the ZipFile is being used by the Java ClassLoader. Maybe you need to investigate the jars on your classpath to see which jars have massive files in them.

**Reply by raboof** [reply]
Indeed it would be good to understand where this `ZipFile$Source` is coming from / created.

This is the Eclipse Memory Analyzer (MAT), right? I'm not sure anymore how to read this - I see the URLJarFile which indeed suggests it'd be a good idea to see if there's any huge jars on the classpath, but it seems a bit odd that the 'Retained Heap' of none of the GC roots seems to be around 4,94,11,544 in size. Also that number seems oddly formatted?

It would be good to understand what this `ZipFile$Source` is used for, indeed.

**Reply by pjfanning** [reply]
@raboof that's how South Asians format numbers - they use [lakhs](https://en.wikipedia.org/wiki/Lakh) and crores.

**Reply by raboof** [reply]
Interesting!

**Reply by Susmit07** [reply]
Exactly guys I am also a bit confused, explicitly not using ZipFile neither we are reading zip data its plain hdfs and s3 connectors reading parquet file, need to see the classpath
